cli/client.go

```diff
@@ -119,7 +125,16 @@ var clientDealCmd = &cli.Command{
 		if err != nil {
 			return err
 		}
-		proposal, err := api.ClientStartDeal(ctx, data, a, miner, types.BigInt(price), uint64(dur))
+
+		ref := &storagemarket.DataRef{
+			TransferType: storagemarket.TTGraphsync,
+			Root:         data,
```

Comment:
It looks like we are still passing a root CID when the deal is created... and I'm not sure that will work. My understanding is for offline the CAR file will be assembled outside of filecoin, and a commP generated from it.

The current code here will still internally generate a CAR file from this root CID (assuming data is in the local blockstore) -- which means we'd be assuming filecoin's CAR writing methods, which may not exactly match the commP of the a CAR file generated elsewhere. It seems like for offline we need to directly pass commP for this to work, and skip the step of generating a car file internally. The relevant line we want to skip somehow is here: https://github.com/filecoin-project/go-fil-markets/blob/master/storagemarket/impl/client_utils.go#L44

Reply:
We should definitely assume a canonical way of generating car file commPs from the input data. If there is a mismatch, we should rectify that.

cc @mikeal

Reply:
It depends on what you mean by “input data.” The same CAR file bytes obviously need to produce the same commP, but there *are* CAR files that **are not** deterministically generated. So you can definitely have situations in which there are two commP’s for two *different* CAR files with the same root CID.

This isn’t an issue in the regular deal flow because two parties have to create the same CAR file, so they will always be deterministically generated. But in the offline deal flow we **do** have CAR files that cannot be deterministically generated from the root CID. We’re creating these for large datasets and some of the CAR files have to be created by pulling the data from source targets in parallel. Deterministic CAR file generation requires that they be created serially to preserve a consistent ordering. We should also expect that in the future people may need to create CAR files for partial graphs and you’d have a similar situation in which two different CAR files have the same root CID.

It’s also just a good idea, given the offline deal flow, to not make assumptions about the determinism of a CAR file based only on the root CID. It’ll always be possible for someone to put a CAR file in the network that is non-deterministic and the only way to validate that it is not deterministic is to parse it the entire CAR file.

Reply:
^^^ @mikeal yes this matches my understanding.